### PR TITLE
Add memcached support.

### PR DESCRIPTION
**Pull request by kareem-wolfssl**
memcached support: add required functions/defines.

Break out session_id_context APIs into separate option WOLFSSL_SESSION_ID_CTX, so they can be used without OPENSSL_EXTRA.

Make wolfSSL_ERR_get_error available for memcached.

Add --enable-memcached.

# Testing

Tested by building with memcached and using its test suite.

### CODE REVIEW

**Review by julek-wolfssl — DISMISSED** [review]

File: configure.ac
Comment (dgarske) [inline configure.ac:5810]: Is `DEBUG_WOLFSSL_VERBOSE` really always required?

File: src/internal.c
Comment (dgarske) [inline src/internal.c:7007]: Nice cleanup on `WOLFSSL_SESSION_ID_CTX`. Thank you

**Review by dgarske — APPROVED** [review]

File: configure.ac
Comment (kareem-wolfssl) [reply configure.ac:5810]: Looks like it was added originally for ERR_print_errors_fp, before I made a dedicated memcached flag HAVE_MEMCACHED.  Now that I have that flag, I will just gate the function ERR_print_errors_fp behind it so memcached doesn't pull in other verbose logging functions.

File: .github/workflows/memcached.yml
Comment (julek-wolfssl) [inline .github/workflows/memcached.yml:47]: Needs to be changed before merging. Leaving comment to block merging until the `osp` PR is merged.

File: src/ssl.c
Comment (julek-wolfssl) [inline src/ssl.c:29503]: Using `ssl->options.connectState` and `ssl->options.acceptState` has been a source of bugs in the past because they mean different things in TLS 1.3 and <= 1.2. All the `*BEGIN` values evaluate to `0` but I would prefer if this API would just use `return ssl->options.handShakeState == NULL_STATE;` instead. This would also not require the need to branch on the connection side.

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]

File: src/ssl.c
Comment (kareem-wolfssl) [reply src/ssl.c]: Updated to fix this.  I did notice the existing API wolfSSL_SSL_in_connect_init is using connect/acceptState as well, does this function need to be fixed?

Comment (julek-wolfssl) [reply src/ssl.c]: Yes `wolfSSL_SSL_in_connect_init` should also be modified to use `ssl->options.handShakeState`. Can you modify it to be a copy of `wolfSSL_SSL_in_init`?

File: .github/workflows/memcached.yml
Comment (julek-wolfssl) [reply .github/workflows/memcached.yml:47]: @kareem-wolfssl I merged the osp PR. Please update this.

Comment (kareem-wolfssl) [reply .github/workflows/memcached.yml:47]: Done

File: src/ssl.c
Comment (kareem-wolfssl) [reply src/ssl.c]: Updated per your request, but are you sure this is correct?
It looks like wolfSSL_in_connect_init is supposed to only return 1 if we are acting as a client and wolfSSL_in_init would return 1.

Comment (julek-wolfssl) [reply src/ssl.c]: Let's revert this change for now and we can address these functions in a separate PR. It should check if we are in progress of handshaking.

Comment (kareem-wolfssl) [reply src/ssl.c]: Done

**Review by julek-wolfssl — APPROVED** [review]

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**
